### ISSUE TITLE
Should private members be inaccessible from objects of the same type?

### ISSUE THREAD

**Issue by NeilGirdhar** [charliermarsh/ruff#3162]
Should a class factory not be able to access the object it creates's private methods?  Should a comparison function not be able to access an object of the same type's private members?
```
from typing import Self, Any

class X:
    _x: int

    def _f(self) -> None:
        ...

    @classmethod
    def factory(cls, x: int) -> Self:
        retval = cls()
        retval._f()  # SLF001 Private member accessed
        return retval

    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, X):
            return False
        return other._x == self._x  # SLF001 Private member accessed
```

**Comment by not-my-profile** [comment]
Yes these are false positives. The problem with this is that ruff isn't a static type checker. Static type checkers understand which type `cls()` returns and they support type narrowing i.e. they understand that the type of `other` is `X` after the above `isinstance` check ... ruff doesn't.

And while we could easily implement some logic to handle the exact cases of your example that logic would fail if e.g. retval was reassigned or the `isinstance` check was more complicated ... properly handling this requires proper static type checking.

**Comment by NeilGirdhar** [comment]
Yes, I considered that it might be difficult.  Shall I close this issue then?

**Comment by charliermarsh** [comment]
Yeah probably hard to resolve right now. Eventually, I hope!
